File: logger/src/main/java/ru/surfstudio/android/logger/logging_strategies/impl/concrete/timber/TimberLoggingStrategy.java
package ru.surfstudio.android.logger.logging_strategies.impl.concrete.timber;

import ru.surfstudio.android.logger.logging_strategies.impl.base.AbstractGroupedLoggingStrategy;
import timber.log.Timber;

public class TimberLoggingStrategy extends AbstractGroupedLoggingStrategy {

    @Override
    protected void log(int priority, Throwable t) {
        setClickableLink();
        Timber.log(priority, t);
    }

    @Override
    protected void log(int priority, String message, Object... args) {
        setClickableLink();
        Timber.log(priority, message, args);
    }

    @Override
    protected void log(int priority, Throwable t, String message, Object... args) {
        setClickableLink();
        Timber.log(priority, t, message, args);
    }

    private static void setClickableLink() {
        StackTraceElement[] stackTrace = new Throwable().getStackTrace();
        if (stackTrace.length < 2) {
            throw new IllegalStateException("Synthetic stacktrace didn't have enough elements: are you using proguard?");
        }
        StackTraceElement element = stackTrace[2];
        String tagMsg = String.format("(%s:%s) ", element.getFileName(), element.getLineNumber());
        Timber.tag(tagMsg);
    }
}

File: template/base_feature/src/main/java/ru/surfstudio/standard/application/logger/strategies/local/TimberLoggingStrategy.kt
package ru.surfstudio.standard.application.logger.strategies.remote.timber

import ru.surfstudio.android.logger.logging_strategies.LoggingStrategy
import timber.log.Timber

/**
 * Logging strategy with [Timber] usage
 */
class TimberLoggingStrategy : LoggingStrategy {

    init {
        Timber.plant(Timber.DebugTree())
    }

    override fun log(priority: Int, t: Throwable?, message: String?, vararg args: Any?) {
        setClickableLink()
        Timber.log(priority, t, message, *args)
    }

    private fun setClickableLink() {
        val stackTrace = Throwable().stackTrace
        if (stackTrace.size < 2) {
            throw IllegalStateException("Synthetic stacktrace didn't have enough elements: are you using proguard?")
        }
        val element = stackTrace[2]
        val tagMsg = String.format("(%s:%s) ", element.fileName, element.lineNumber)
        Timber.tag(tagMsg)
    }
}
